### DISCUSSION TITLE
[Detector Support]: TensorRT detects motion but not objects

### DISCUSSION THREAD

**Discussion by jakeek** [blakeblackshear/frigate#11142] [Detector Support] [answered]
### Describe the problem you are having

Hello, I have been working on setting up TensorRT over the past couple days but have hit a bit of snag with object detection. It is detecting motion fine but will not identify objects whatsoever. Switching back to the CPU detector resolves the issue and objects are detected again. I think I've read just about every related troubleshooting post on the internet. Thanks in advance!

### Version

0.13.2-6476f8a

### Frigate config file

```yaml
logger:
  logs:
    frigate.events: debug

mqtt:
  host: 192.168.17.36
  port: 1883
  user: "user"
  password: "pass"
  topic_prefix: "frigate"

timestamp_style:
  position: tl
  format: "%m/%d/%Y %H:%M:%S"
  thickness: 2
  effect: solid

go2rtc:
  streams:
    cam:
      - rtsp://user:pass@192.168.99.20:554/cam/realmonitor?channel=1&subtype=0
    cam_sub:
      - rtsp://user:pass@192.168.99.20:554/cam/realmonitor?channel=1&subtype=1

birdseye:
  enabled: true
  mode: continuous

detectors:
  tensorrt:
    type: tensorrt
    device: 0
#  cpu:
#    type: cpu 

model:
  path: /config/model_cache/tensorrt/yolov4-tiny-416.trt
  labelmap_path: /labelmap.txt
  input_tensor: nhwc
  input_pixel_format: rgb
  width: 416
  height: 416

cameras:
  cam:
    enabled: true
    ffmpeg:
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://192.168.17.34:8554/cam
          hwaccel_args: preset-nvidia-h264
          input_args: preset-rtsp-restream-low-latency
          roles:
            - record
        - path: rtsp://192.168.17.34:8554/cam_sub
          hwaccel_args: preset-nvidia-h264
          input_args: preset-rtsp-restream-low-latency
          roles:
            - detect
    record:
      enabled: true
      retain:
        days: 5
        mode: all
      events:
        retain:
          default: 30
          mode: motion
    detect:
      enabled: true
      width: 704
      height: 480
      fps: 5
    snapshots:
      enabled: True
      bounding_box: True
      retain:
        default: 30
    objects:
      track:
        - person
        - cat
        - dog
        - car
        - motorcycle
        - bicycle
```

### docker-compose file or Docker CLI command

```yaml
version: "3.8"
services:
  frigate:
    container_name: frigate
    restart: unless-stopped
    image: ghcr.io/blakeblackshear/frigate:stable-tensorrt
    deploy:    # <------------- Add this section
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]
    shm_size: "64mb"
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/New_York
      - FRIGATE_RTSP_PASSWORD="pass"
      - USE_FP16=false
      - YOLO_MODELS=yolov3-tiny-416,yolov3-608,yolov4-tiny-416,yolov4-608,yolov7-tiny-416,yolov7-640
 volumes:
      - /etc/localtime:/etc/localtime:ro
      - /opt/appdata/frigate/config:/config
      - "/mnt/NVR/frigate:/media/frigate"
      - type: tmpfs
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "5000:5000"
      - "8554:8554" # RTSP feeds
      - "8555:8555/tcp" # WebRTC over tcp
      - "8555:8555/udp" # WebRTC over udp
    networks:
      - rp-net

networks:
  rp-net:
    external: true
```

### Relevant log output

```shell
s6-rc: info: service s6rc-fdholder: starting
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service s6rc-fdholder successfully started
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service trt-model-prepare: starting
s6-rc: info: service log-prepare: starting
s6-rc: info: service log-prepare successfully started
s6-rc: info: service nginx-log: starting
s6-rc: info: service go2rtc-log: starting
s6-rc: info: service frigate-log: starting
Generating the following TRT Models: yolov7-320,yolov7-tiny-416
Downloading yolo weights
s6-rc: info: service nginx-log successfully started
s6-rc: info: service frigate-log successfully started
s6-rc: info: service go2rtc-log successfully started
s6-rc: info: service go2rtc: starting
s6-rc: info: service go2rtc successfully started
s6-rc: info: service go2rtc-healthcheck: starting
s6-rc: info: service go2rtc-healthcheck successfully started
2024-04-28 22:17:07.059449836  [INFO] Preparing new go2rtc config...
2024-04-28 22:17:07.456366453  [WARN] Using go2rtc binary from '/config/go2rtc' instead of the embedded one
2024-04-28 22:17:07.456375389  [INFO] Starting go2rtc...
2024-04-28 22:17:07.550254540  22:17:07.550 INF go2rtc version 1.8.5 linux/amd64
2024-04-28 22:17:07.550864806  22:17:07.550 INF [api] listen addr=:1984
2024-04-28 22:17:07.550964072  22:17:07.550 INF [rtsp] listen addr=:8554
2024-04-28 22:17:07.551712819  22:17:07.551 INF [webrtc] listen addr=:8555

Creating yolov7-tiny-416.cfg and yolov7-tiny-416.weights
Creating yolov7-320.cfg and yolov7-320.weights

Done.
2024-04-28 22:17:17.056299680  [INFO] Starting go2rtc healthcheck service...

Generating yolov7-320.trt. This may take a few minutes.

/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:147: DeprecationWarning: Use network created with NetworkDefinitionCreationFlag::EXPLICIT_BATCH flag instead.
  builder.max_batch_size = MAX_BATCH_SIZE
/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:149: DeprecationWarning: Use set_memory_pool_limit instead.
  config.max_workspace_size = 1 << 30
/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:172: DeprecationWarning: Use build_serialized_network instead.
  engine = builder.build_engine(network, config)
Generated yolov7-320.trt in 133 seconds

Generating yolov7-tiny-416.trt. This may take a few minutes.

/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:147: DeprecationWarning: Use network created with NetworkDefinitionCreationFlag::EXPLICIT_BATCH flag instead.
  builder.max_batch_size = MAX_BATCH_SIZE
/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:149: DeprecationWarning: Use set_memory_pool_limit instead.
  config.max_workspace_size = 1 << 30
/usr/local/src/tensorrt_demos/yolo/onnx_to_tensorrt.py:172: DeprecationWarning: Use build_serialized_network instead.
  engine = builder.build_engine(network, config)
Generated yolov7-tiny-416.trt in 106 seconds
Available tensorrt models:
yolov7-320.trt
yolov7-tiny-416.trt
s6-rc: info: service trt-model-prepare successfully started
s6-rc: info: service frigate: starting
s6-rc: info: service frigate successfully started
s6-rc: info: service nginx: starting
s6-rc: info: service nginx successfully started
s6-rc: info: service legacy-services: starting
2024-04-28 22:21:21.456783645  [INFO] Preparing Frigate...
2024-04-28 22:21:21.458382098  [INFO] Starting NGINX...
s6-rc: info: service legacy-services successfully started
2024-04-28 22:21:21.471191897  [INFO] Starting Frigate...
2024-04-28 22:21:22.006128263  2024/04/28 22:21:22 [error] 366#366: *1 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.17.36, server: , request: "GET /api/stats HTTP/1.1", upstream: "http://127.0.0.1:5001stats", host: "192.168.17.34:5000"
2024-04-28 22:21:23.997762255  [2024-04-28 22:21:23] frigate.app                    INFO    : Starting Frigate (0.13.2-6476f8a)
2024-04-28 22:21:24.527889107  [2024-04-28 22:21:24] peewee_migrate.logs            INFO    : Starting migrations
2024-04-28 22:21:24.536507661  [2024-04-28 22:21:24] peewee_migrate.logs            INFO    : There is nothing to migrate
2024-04-28 22:21:24.544710559  [2024-04-28 22:21:24] frigate.app                    INFO    : Recording process started: 650
2024-04-28 22:21:24.552703047  [2024-04-28 22:21:24] frigate.app                    INFO    : go2rtc process pid: 100
2024-04-28 22:21:24.603788087  [2024-04-28 22:21:24] frigate.app                    INFO    : Output process started: 662
2024-04-28 22:21:24.614325195  [2024-04-28 22:21:24] detector.tensorrt              INFO    : Starting detection process: 660
2024-04-28 22:21:24.629843515  [2024-04-28 22:21:24] frigate.app                    INFO    : Camera processor started for test: 669
2024-04-28 22:21:24.637213954  [2024-04-28 22:21:24] frigate.app                    INFO    : Capture process started for test: 671
2024-04-28 22:21:24.867127746  [2024-04-28 22:21:24] frigate.detectors.plugins.tensorrt INFO    : Loaded engine size: 174 MiB
2024-04-28 22:21:25.118602850  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +6, GPU +8, now: CPU 296, GPU 252 (MiB)
2024-04-28 22:21:25.128034099  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] Init cuDNN: CPU +1, GPU +10, now: CPU 297, GPU 262 (MiB)
2024-04-28 22:21:25.140337724  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +174, now: CPU 0, GPU 174 (MiB)
2024-04-28 22:21:25.140574864  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +0, GPU +8, now: CPU 123, GPU 254 (MiB)
2024-04-28 22:21:25.140798090  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] Init cuDNN: CPU +0, GPU +8, now: CPU 123, GPU 262 (MiB)
2024-04-28 22:21:25.142478387  [2024-04-28 22:21:25] frigate.detectors.plugins.tensorrt INFO    : [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +29, now: CPU 0, GPU 203 (MiB)
2024-04-28 22:21:32.882998357  [2024-04-28 22:21:32] frigate.events.maintainer      DEBUG   : Event received: tracked_object start test 1714357292.697738-aj980d
2024-04-28 22:21:33.548426239  [2024-04-28 22:21:33] frigate.events.maintainer      DEBUG   : Event received: tracked_object end test 1714357292.697738-aj980d
```

### Operating system

Debian

### Install method

Docker Compose

### Coral version

Other

### Any other information that may be helpful

I have deleted model folder and rebuilt several times.
GPU is a 1060 6GB
Driver version 535.161.08
CUDA Version 12.2

**Comment by NickM-27** [comment]
I might suggest starting with using yolov7-320 and see what that looks like

**Comment by jakeek** [comment]
> I might suggest starting with using yolov7-320 and see what that looks like

Yep, that is the one I started with as the default but had no luck. I began trying a few of the different options to see if it was the model causing the issue and picked a couple options from each YOLO version, which is what is currently in my docker compose yaml. So far, no differences in regards to object detection regardless of the model I select. However, a few of the models cause my frigate container to restart periodically.

**Reply by NickM-27** [reply]
sounds like something is setup incorrectly still, try removing `  labelmap_path: /labelmap.txt`

**Reply by jakeek** [reply]
I added that option in relatively recently in my troubleshooting to see if that list wasn't being found for whatever reason. So I don't think it's the problem. However, it's removed and the issue persists.

**Reply by NickM-27** [reply]
very odd, can we see a screenshot of the system page with the tensorrt running and motion being detected

**Reply by jakeek** [reply]
![frigate_system_page_motion](https://github.com/blakeblackshear/frigate/assets/41632114/79463d3e-9246-467b-bdd7-71c030b5e4f0)
Here you go! Just FYI, switched back to default yolov7-320 for this in case it's relevant at some point.

**Reply by NickM-27** [reply]
so the fact that we can see an inference speed means that inferences are being run successfully. I would suggest watcing the debug live view with motion, region, and bounding boxes enabled and seeing what that looks like. a screen video might be helpful to see

**Reply by jakeek** [reply]
https://github.com/blakeblackshear/frigate/assets/41632114/e557fa1b-485f-4738-90f8-b95663540203

Video attached!

**Reply by NickM-27** [reply]
this is a difficult angle and might just be that person is not scoring well. You might try running on a demo video like https://demo.frigate.video/api/events/1714356026.240527-xuy7xj/clip.mp4?download=true with a camera config liek https://docs.frigate.video/development/contributing#2-modify-your-local-config-file-for-testing

**Reply by jakeek** [reply]
https://github.com/blakeblackshear/frigate/assets/41632114/8d2a323b-13d6-4152-b572-1da7cc486d50

Still the same result unfortunately.

Frigate config:
cameras:
  test:
    ffmpeg:
      inputs:
        - path: /config/clip_crosswalk_1714356021.240527-1714356046.146615.mp4
          input_args: -re -stream_loop -1 -fflags +genpts
          roles:
            - detect
      detect:
      enabled: true
      width: 1280
      height: 720
      fps: 10
   objects:
      track:
        - person
        - cat
        - dog
        - car
        - motorcycle
        - bicycle
 
(sorry for the formatting)

**Reply by NickM-27** [reply]
maybe try deleting the /config/model_cache folder and let the models redownload. I can't tell why this would be occurring

**Reply by jakeek** [reply]
Model cache has been deleted and regenerated, still same result. I also added in my logs to the top in case you can see something that I can't.

**Reply by NickM-27** [reply]
maybe @NateMeyer has some ideas, I don't have any other ones at the moment

**Reply by jakeek** [reply]
Appreciate the help with this @NickM-27 !

**Reply by jakeek** [reply]
@NickM-27 I have a few updates. I updated my nvidia drivers to 550 and cuda to 12.4. This seems to have _slightly_ helped, but we're finally seeing object detection which is neat! However, it's still acting a little strange, I'll post videos. For reference, I'm using the yolov7-320 model. First video is where objects are detected on the test video feed. It's weird because my detector settings do not match the video feed (I forgot to switch from my security cam settings) but it works...

https://github.com/blakeblackshear/frigate/assets/41632114/ffeedc58-bc22-4c0f-b85d-842e018391c8

When I switch the detector settings to match the video feed resolution I lose object detection again as seen in the following video:

https://github.com/blakeblackshear/frigate/assets/41632114/11402031-c98e-4dee-a877-c404bb62c866

So yeah a little bit of progress but I'm still stumped. Even in the situation where object detection is barely working it's not exactly behaving as I would expect. It seems to pick the person up extremely late and misses the car entirely. Whereas on my CPU detector things are detected seemingly instantly.

**Reply by NickM-27** [reply]
> When I switch the detector settings to match the video feed resolution I lose object detection again as seen in the following video:

what do you mean by this? you should not be changing detector settings based on any video. The detector settings should match the model used

**Reply by jakeek** [reply, marked as answer]
I might be using the terminology wrong...but I was under the assumption based on the docs that the "detect:" section needed to match the camera feed resolution used for detection.

More importantly though, I figured out what was wrong. Dumb mistake on my part. Under "model:" I set "input_tensor: nhwc" rather than the correct "input_tensor: nchw". I changed it to nchw and it works perfectly now.

https://github.com/blakeblackshear/frigate/assets/41632114/531e572b-b887-464a-b706-2c374edc54b6

Thanks again for your help @NickM-27
